### PR TITLE
Mark internalUser with operator privileges

### PR DESCRIPTION
**Pull request by pgomulka**
In serverless where non operator users are not able to set operator-only settings we want to send transport requests using internaluser. In order to create any resources (indices, templates) using non-public settings an internalUser has to have operator privileges

### CODE REVIEW

**Comment by gwbrown** [comment]
@elasticmachine update branch

**Comment by ywangd** [comment]
@pgomulka I don't think the proposed change is necessary because
1. We already have special handling to exempt internal users from operator privileges [tagging](https://github.com/elastic/elasticsearch/blob/42fbe0b1dcc3dfce4e8976f618110b227078916b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/operator/OperatorPrivileges.java#L103-L106) and [checking](https://github.com/elastic/elasticsearch/blob/42fbe0b1dcc3dfce4e8976f618110b227078916b/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/operator/OperatorPrivileges.java#L131-L133), which effectively makes all internal users operators.
2. You said this is for serverless. In serverless, currently operator privileges are enforced only at REST layer and internal users do not work at REST layer.

Please let me know if the above makes sense. Thanks!

**Comment by pgomulka** [comment]
@ywangd 
> You said this is for serverless. In serverless, currently operator privileges are enforced only at REST layer and internal users do not work at REST layer.

I am working on a public/non-public settings for serverless. The infrastructure for public/non-public settings is enforced in a transport layer https://github.com/elastic/elasticsearch-serverless/pull/478 
I am expecting that we might want system indices to use non-public settings (this is the case at the moment, but perhaps some of the index templates will be removed /replaced in serverless)

**Review by ywangd — COMMENTED** [review]
I am a little bit unsure about this change because:
1. Strictly speaking, internal users are not operator users, e.g. they are not designated in the operator_user.yml file. Maybe they should be marked with a different category value instead of `operator`, e.g. `_security_privilege_category: internal_user`. 
2. Marking an internal user is actually redundant because we can already tell it is intenral by inspecting the authentication object. So what you really need is just a _method_ to tell you whether the current user is operator like. Intenrally this method can check whether it is marked as `operator` and whether it is an internal user, which are implementation details.
3. We previously talked about providing such method so it can be called by other parts of the code. But I remember we had concerns about proliferating calls to the method in different parts of the code can make it hard to reason about. I generally prefer security checks to be performed within security related code.

@jakelandis Could you please shed some light on this change? It might be possible that I am overthinking. Thanks!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/operator/OperatorPrivileges.java
Comment (jakelandis) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/operator/OperatorPrivileges.java:112]: Discussed with @pgomulka in person. The act of adding this here isn't too much of a concern, all internal users should have full access to roles with out operator privs kicking in and it would be kinda silly to externalize our internal users into a config file.  At a high level the only concern by adding here is that it means it read elsewhere to implement operator privs, but that is already a concern due to the public global nature of thread context... we are just extending that potential to internal users that already exists for external users. 

We discussed a different PR that reads this header value but in a different location than here in this class. That usecase is why this is being added to this class and there are many tradeoffs to consider. The largest tradeoff is that operators privs, for that use case, will not be done via the operator registry. Rather that use case will use action filters to manipulate the responses if users are not operators effectively splitting where operator privs are applied, hence the need to read the thread context in multiple location (and the need to set the context for internal users). Those tradeoffs had already been discussed and accepted (I was not directly part of those). I only have a second hand understanding of those discussions and tradeoffs, but it that seems like an acceptable decision given circumstances and that decision necessitates this change. However, I did encourage that any strategy that implements operator privs outside of the registry have some association back to the registry so there is at least some bread crumbs that allow a dev to know where to look for non-operator related behavior. 

also ++ to remove from the checking side.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/operator/DefaultOperatorPrivilegesTests.java
Comment (jakelandis) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/operator/DefaultOperatorPrivilegesTests.java:154]: please remove commented code.

**Review by jakelandis — CHANGES_REQUESTED** [review]
Will approve once checking side is removed and test is adjusted per comments.

**Comment by jakelandis** [comment]
> Strictly speaking, internal users are not operator users, e.g. they are not designated in the operator_user.yml file. Maybe they should be marked with a different category value instead of operator, e.g. _security_privilege_category: internal_user.

I am fine with calling internal users operators. I see operator privs as way to identify non-operators and treat them special and we don't ever need to treat internal users special in that respect. 

>    Marking an internal user is actually redundant because we can already tell it is intenral by inspecting the authentication object. So what you really need is just a method to tell you whether the current user is operator like. Intenrally this method can check whether it is marked as operator and whether it is an internal user, which are implementation details.

I generally agree. As [commented](https://github.com/elastic/elasticsearch/pull/97569#discussion_r1286173543) above this is needed since it being read elsewhere. 


>    We previously talked about providing such method so it can be called by other parts of the code. But I remember we had concerns about proliferating calls to the method in different parts of the code can make it hard to reason about. I generally prefer security checks to be performed within security related code.

I generally agree. As [commented](https://github.com/elastic/elasticsearch/pull/97569#discussion_r1286173543) above there are tradeoff for a design not seen in this PR that necessitate this change.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/operator/OperatorPrivileges.java
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/operator/OperatorPrivileges.java:103]: We should remove this comment as well.

**Comment by ywangd** [comment]
Thanks Jake for the insights. I am onboard with the changes and will leave it to Jake for approval. Thanks!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/operator/OperatorPrivileges.java
Comment (pgomulka) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/operator/OperatorPrivileges.java]: but isn't it still partially relevant?
I changed the first part to `Let internal users pass, and mark him as an operator`

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/operator/OperatorPrivileges.java
Comment (jakelandis) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/operator/OperatorPrivileges.java:139]:
```suggestion
```

**Review by jakelandis — APPROVED** [review]
LGTM (with a couple nit pick suggestions around comments)

**Comment by pgomulka** [comment]
@elasticmachine update branch

**Pull request merged**
